### PR TITLE
Edit drawable key

### LINKED ISSUE #208: Allow for changing drawable keys

### LINKED ISSUE #215: Modularize dk-input + save button into its own component
Entity name, collision type creation, and map creation all use a similar pattern of an input field follow by a button to the right to save the value. This should be moved into its own component.

### PR DESCRIPTION
**Pull request by tlein**
Resolves #208 
Resolves #215

### CODE REVIEW

File: src/duckling/game/drawable/generic-drawable.component.ts [hunk 1]
```diff
@@ -77,4 +91,14 @@ export class GenericDrawableComponent {
     onInactivePressed(inactive : boolean) {
         this.drawableChanged.emit(immutableAssign(this.drawable, {inactive: inactive}));
     }
+
+    get keyValidator() : Validator {
```
Comment (tlein) [inline src/duckling/game/drawable/generic-drawable.component.ts]: should check for duplicate keys

Comment (tlein) [reply src/duckling/game/drawable/generic-drawable.component.ts]: fixed

File: src/duckling/controls/delayed-save-input.component.ts [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+    OnChanges,
+    OnInit
+} from '@angular/core';
+
+import {Validator} from './validated-input.component';
+
+import {DeleteButtonComponent, InputComponent} from '../controls';
+
+@Component({
+    selector: "dk-delayed-save-input",
```
Comment (jeffswenson) [inline src/duckling/controls/delayed-save-input.component.ts]: Maybe name it dk-edit-input?

File: src/duckling/game/drawable/drawable-attribute.component.ts [hunk 3]
```diff
@@ -30,4 +32,15 @@ export class DrawableAttributeComponent {
     onDrawableChanged(newDrawable : Drawable) {
         this.attributeChanged.emit(immutableAssign(this.attribute, {topDrawable: newDrawable}));
     }
+
+    get drawableKeyValidator() : Validator {
+        return (value : string) => {
+            if (!value || value === "") {
```
Comment (jeffswenson) [inline src/duckling/game/drawable/drawable-attribute.component.ts]: "" is falsey

File: src/duckling/game/drawable/drawable.ts [hunk 4]
```diff
@@ -65,4 +67,33 @@ export let defaultDrawable : Drawable = {
...
+            return null;
+        case DrawableType.Animated:
+            for (let drawable of (parentDrawable as AnimatedDrawable).frames) {
+                let childByKey = getDrawableByKey(drawable, key);
+                if (childByKey) {
+                    return childByKey;
+                }
+            }
+            return null;
+        case DrawableType.Shape:
```
Comment (jeffswenson) [inline src/duckling/game/drawable/drawable.ts]: This should be the default case

File: src/duckling/game/drawable/drawable.ts [hunk 5]
```diff
@@ -65,4 +67,33 @@ export let defaultDrawable : Drawable = {
...
+};
+
+export function getDrawableByKey(parentDrawable : Drawable, key : string) : Drawable {
+    if (key === parentDrawable.key) {
+        return parentDrawable;
+    }
+
+    switch (cppTypeToDrawableType(parentDrawable.__cpp_type)) {
+        case DrawableType.Container:
+            for (let drawable of (parentDrawable as ContainerDrawable).drawables) {
```
Comment (jeffswenson) [inline src/duckling/game/drawable/drawable.ts]: return parentDrawable.drawables.Any(drawable => getDrawableByKey(drawable, key));

File: src/duckling/controls/delayed-save-input.component.ts (hunk 2 above)
Comment (tlein) [reply src/duckling/controls/delayed-save-input.component.ts]: fixed

File: src/duckling/game/drawable/drawable-attribute.component.ts (hunk 3 above)
Comment (tlein) [reply src/duckling/game/drawable/drawable-attribute.component.ts]: fixed

File: src/duckling/game/drawable/drawable.ts (hunk 4 above)
Comment (tlein) [reply src/duckling/game/drawable/drawable.ts]: fixed

**Pull request merged**